rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,85 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare bytes of bytecode file serialized at HEAD
+with existing [backward compatibility bytecode test files](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc).
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing a file at
+head and comparing it to a known-good file serialized at the previous release.
+This approach operates on the assumption that at HEAD we are able to produce
+a byte-identical serialized artifact, excluding custom header info and debug
+locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-opt %s | stablehlo-translate --serialize --target=0.X.0)
+```
+
+Passing through `stablehlo-opt` in this example removes debug info. An option
+to `stablehlo-translate` can be added to strip debug info and header info. This
+test should be run in CI on all PRs to prevent merging incompatible changes in
+StableHLO (_R1_) and LLVM revision bumps (_R2_).
+
+The StableHLO process for [Contributing Incompatible Changes](https://github.com/openxla/stablehlo/blob/main/docs/vhlo.md#add-versioned-serialization-test)
+requires a versioned test be serialized for each compatibility breaking change,
+which means that we can test forward compatibility between HEAD and each minor
+release of StableHLO. Given that all changes to the StableHLO opset, as well as
+changes to serialization machinery like targeting a newer version of the MLIR
+bytecode format, require bumping the minor version, there is a low chance of
```

Comment:
Does this process leave any chance for forward incompatibilities?

Reply:
FileA: Serialized@0.x.0

Release 0.x.0 - Can serialize a file to the bytes of FileA, can read a file with the bytes of FileA
Release 0.x.1 - Can serialize a file to FileA, can read a file with the bytes of FileA
Release 0.x.2 - Can serialize a file to FileA, can read a file with the bytes of FileA
...

Assuming that a release never broke backward compatibility (can read FileA), and its successor release doesn't break forward compatibility (can generate FileA), then there should be no forward compatibility issues. Given that there are no opset changes or bytecode changes during patch releases, targeting minor versions is preferable and safe (and the ability to target patch versions should possibly be removed).

Reply:
I'm not sure I understand the difference between "Can serialize a file to the bytes of FileA" and "Can serialize a file to FileA". Also, can you update the text of the RFC to elaborate on what "a low chance" means exactly?

Reply:
@burmako let me know if this makes sense now and can be resolved.

Reply:
Okay, let's open a ticket to prohibit targeting patch versions and then this comment can be resolved. After that, let's implement this ticket with very high priority.

Reply:
Created https://github.com/openxla/stablehlo/issues/1653